// @slack/client/index.js
var events = require('./lib/clients/events');
var XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;

var scores = null;

function compareNombres(a, b) {
  return b - a;
}

function readTextFile(file) {
  var rawFile = new XMLHttpRequest();
  rawFile.open("GET", file, false);
  rawFile.onreadystatechange = function ()
  {
    if(rawFile.readyState === 4)
    {
      if(rawFile.status === 200 || rawFile.status == 0)
      {
        scores = JSON.parse(rawFile.responseText);
      }
    }
  }
  rawFile.send(null);
}

readTextFile("test.json");


module.exports = {
  WebClient: require('./lib/clients/web/client'),
  RtmClient: require('./lib/clients/rtm/client'),
  LegacyRtmClient: require('./lib/clients/default/legacy-rtm'),
  CLIENT_EVENTS: {
    WEB: events.CLIENT_EVENTS.WEB,
    RTM: events.CLIENT_EVENTS.RTM
  },
  RTM_EVENTS: events.RTM_EVENTS,
  RTM_MESSAGE_SUBTYPES: events.RTM_MESSAGE_SUBTYPES,
  MemoryDataStore: require('./lib/data-store/memory-data-store')
};
var RtmClient = require('./lib/clients/rtm/client');

var token = 'xoxb-25294436146-IsORJSbMQGIswHkscmGJ2Gy2';

var rtm = new RtmClient(token);

var reponse = null;

rtm.start();

var CLIENT_EVENTS = require('@slack/client').CLIENT_EVENTS;

rtm.on(CLIENT_EVENTS.RTM.AUTHENTICATED, function (rtmStartData) {
  console.log("bien loggé.");

});

rtm.on('message', function handleRtmMessage(message) {
  console.log("message recu !")
if (message.text!=undefined) {
  if (message.text.substr(0, 7) === "!manger") {
    var restau = ["Corse", "Subway", "Mongoo", "Pizza", "Japonais", "Japonais jusqu'à 15 heures", "MACDO", "Pegast", "les burgers Hipster bio", "Bagel", "Toto pizza", "Goutu (les sandwich entre 1 et 3e)", "un nouveau resto tu testeras"];
    var res = Math.trunc(Math.random() * restau.length);
    rtm.sendMessage("Aujourd'hui vous allez manger " + restau[res] + ". Bon appétit à tous !", message.channel, function messageSent() {
      console.log("Requête de restaurant effectuée et traitée. l'index " + res + " à été selectionné");
    });
  }
  else if (message.text.substr(0, 9) === "!shutdown" && message.user === "U0M2Z0B2N") {
    rtm.sendMessage("Bonne nuit les petits !", message.channel, function messageSent() {
    });
    console.log(scores);
    var rendu = "";
    console.log(Object.keys(scores));
    for (var index in Object.keys(scores)) {
      ID2 = Object.keys(scores)[index];
      console.log(ID);
      rendu = rendu + "<@" + ID2 + "> totalise " + scores[ID2] + " points.\n";
    }
    rtm.disconnect();
  }
  else if (message.text.indexOf("salut") > -1) {
    rtm.sendMessage("Salut <@" + message.user + "> !!!", message.channel, function messageSent() {
    });
  }
  else if (message.text.substr(0, 5) === "!user") {
    rtm.sendMessage("Salut!" + message.user, message.channel, function messageSent() {
    });
  }
  else if (message.text.substr(0, 5) === "!wizz") {
    var user = message.text.substr(6);
    rtm.sendMessage("Salut " + user + " !!!", message.channel, function messageSent() {
    });
  }
  else if (message.text === "!PlusMoins") {
    reponse = Math.trunc(Math.random() * 1000);
    console.log("jeu du plus ou moins commencé ! réponse = " + reponse);
    rtm.sendMessage("Vous allez jouer au plus ou moins.\nJ'ai choisi un entier entre 0 et 1000 inclus. \nTapez '!plusmoins ' suivi d'un chiffre pour proposer une réponse.", message.channel, function messageSent() {
    });
  }
  else if (message.text.substr(0, 10) === "!plusmoins" && reponse != null) {
    var proposition = parseInt(message.text.substr(11));
    if (proposition == reponse) {
      reponse = null;
      rtm.sendMessage("Bravo <@" + message.user + ">. Tu as gagné.", message.channel, function messageSent() {
      });
    } else if (proposition < reponse) {
      rtm.sendMessage("C'est plus !", message.channel, function messageSent() {
      });
    } else if (proposition > reponse) {
      rtm.sendMessage("C'est moins !", message.channel, function messageSent() {
      });
    }
  }
  else if (message.text === "!score") {
    console.log(scores);
    var ID = message.user;
    rtm.sendMessage("Bravo <@" + ID + "> ! Tu as " + scores[ID][0] + " points !!!", message.channel, function messageSent() {
    });
  }
  else if (message.text === "!scores") {
    console.log(scores);
    var rendu = "";
    var points = []
    for (var index in Object.keys(scores)) {
      ID2 = Object.keys(scores)[index];
      points.push(scores[ID2][0]);
    }
    console.log("scores non triés : "+points);
    points.sort(compareNombres);
    console.log("scores triés : "+points);
    for (var index in points) {
      var temp = points[index];
      console.log("points : "+temp);
      var temp2=scores;
      listeID = Object.keys(temp2);
      for (var index2 in listeID) {
        console.log("index2 : "+index2);
        var ID = listeID[index2];
        console.log("ID : "+ID);
        console.log("Points et TS de cette personne : "+temp2[ID]);
        if (temp==temp2[ID][0]){
          rendu = rendu + "<@" + ID + "> totalise " + temp2[ID][0] + " points.\n";
          //delete temp2[ID];
        }
      }
    }
    console.log(rendu);
    rtm.sendMessage(rendu, message.channel, function messageSent() {
      console.log("On a bien renvoyé les scores triés !");
    });
    console.log("Scores Totaux : "+Object.keys(scores).length);
  }
  else if (message.text === "!jdh") {
    var points = Math.trunc(Math.random() * 110)-10;
    var ID = message.user;
    if (Object.keys(scores).indexOf(ID) == -1) {
      rtm.sendMessage("Bravo <@" + ID + "> ! Tu marques " + points + " points !!!", message.channel, function messageSent() {
      });
      console.log("On a pas encore cette personne.");
      scores[ID] = [points, Math.trunc(message.ts)];
    } else if (message.ts - scores[ID][1] < 3600) {
      var ts = Math.trunc(3600 - message.ts + scores[ID][1]);
      var minutes = Math.trunc(ts / 60);
      var secondes = Math.trunc(ts % 60);
      rtm.sendMessage("Tu ne peux pas encore rejouer. tu pourras rejouer dans : " + minutes + " minutes et " + secondes + " secondes.", message.channel, function messageSent() {
      });
    }
    else {
      rtm.sendMessage("Bravo <@" + ID + "> ! Tu marques " + points + " points !!!", message.channel, function messageSent() {
      });
      console.log(scores[ID]);
      scores[ID][0] = scores[ID][0] + points;
      scores[ID][1] = Math.trunc(message.ts);
      console.log(scores[ID]);
    }
    console.log(scores);
  }
}
});


//victor = U0M2ZCL6T
//alexandre = U0M30CPEJ
//hermance = U0M39GS4V
//abensimon = U0PLYKW9E
//raphael = U0M2X0W3V
//harold = U0M307X16

//sacha = U0M2Z0B2N
